reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java

```diff
@@ -29,12 +29,12 @@
   static void parse(Class<?> moduleClass, @Nullable Object instance, Scope.Builder scopeBuilder) {
     for (Class<?> target : Reflection.getDistinctTypeHierarchy(moduleClass)) {
       for (Method method : target.getDeclaredMethods()) {
-        if ((method.getModifiers() & PRIVATE) != 0) {
-          throw new IllegalArgumentException("Private module methods are not allowed: " + method);
+        Annotation[] annotations = method.getAnnotations();
+        if ((method.getModifiers() & PRIVATE) != 0 && (annotations == null)) {
```

Comment:
Hmm I'm not sure this check is suitable since if it were written in Kotlin you'd presumably get a `@NonNull` annotation. Ideally we only check when calling `addBinding` which is where we absolutely know you're trying to add a Dagger method.

Or, optionally, maybe we just don't validate this and let the real compiler catch it?

Reply:
Let me take a swing at moving the check in to `addBinding`.

Reply:
Since @Binds and @BindsOptionalOf must be abstract methods anyways, they can't be private, so I just added the check right before addBinds for `@Provides` annotated methods.

Reply:
Works for me!